packages/data/src/components/use-select/index.js

```diff
@@ -191,10 +191,17 @@ export default function useSelect( mapSelect, deps ) {
 		latestRegistry.current = registry;
 		latestMapSelect.current = _mapSelect;
 		latestIsAsync.current = isAsync;
-		latestMapOutput.current = mapOutput;
 		latestMapOutputError.current = undefined;
 	} );
 
+	useIsomorphicLayoutEffect( () => {
+		if ( ! hasMappingFunction ) {
+			return;
+		}
+
+		latestMapOutput.current = mapOutput;
+	}, [ mapOutput ] );
+
```

Comment:
A little more elegant fix would be to keep the `latestMapOutput.current` write in the original hook, but do it only if the selector was already called:
```js
let selectorRan = false;
if ( conditions ) {
  mapOutput = wrapSelect( mapSelect );
  selectorRan = true;
}

useIsomorphicLayoutEffect( () => {
  if ( selectorRan ) {
    latestMapOutput.current = mapOutput;
  }
  // ... other writes in this hook stay intact
} );
```
I tried it locally and all `useSelect` unit tests are passing. Let's see how the entire Gutenberg unit and e2e suite will do.

Also, we should the same changes in `useSuspenseSelect`. Its implementation is similar but slightly different, and we didn't unify them together yet.